Migrate promote job to Python

vfreex opened the pull request:
Created `promote-assembly` job to replace the existing `promote` and
`multi_promote` jobs.
See `jobs/build/promote-assembly/README.md` for instructions.

joepvd commented on jobs/build/promote-assembly/Jenkinsfile:
```diff
@@ -0,0 +1,344 @@
+            parameters: [
+                buildlib.param('String', 'FROM_RELEASE_TAG', from_releases.join(",")),
+                buildlib.param('String', 'RELEASE_NAME', release_info.name),
+                buildlib.param('String', 'ADVISORY_NUM', release_info.advisory?: 0),
```
Cast to string

vfreex replied:
done

joepvd commented on jobs/build/promote-assembly/Jenkinsfile:
```diff
@@ -0,0 +1,344 @@
+                buildlib.param('String', 'FROM_RELEASE_TAG', from_releases.join(",")),
+                buildlib.param('String', 'RELEASE_NAME', release_info.name),
+                buildlib.param('String', 'ADVISORY_NUM', release_info.advisory?: 0),
+                booleanParam(name: 'CANDIDATE_CHANNEL_ONLY', value: true),
```
Job does not have this param anymore

vfreex replied:
Will remove it.

joepvd commented on jobs/build/promote-assembly/Jenkinsfile:
```diff
@@ -0,0 +1,358 @@
+    def release_info = {}
+
+    stage("promote release") {
+        sh "mkdir -p ./artcd_working"
```
As a defensive action, could a `rm -rf ./artcd_working` be executed?

vfreex replied:
Yes, that's a good idea.

joepvd commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+            raise
+        return release_infos
+
+    async def promote_arch(self, release_name: str, arch: str, previous_list: List[str], metadata: Optional[Dict], reference_release: Optional[str], tag_stable: bool):
```
I'd love to see tests for this method. Which is not a merge blocker, but let's file a card.

vfreex replied:
😆 I intended to add unit tests for all methods in the pipeline class, but I haven't done that yet. Let me pick the most important ones and add them first.

joepvd commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+        brew_arch = util.brew_arch_for_go_arch(arch)  # ensure we are using Brew arches (e.g. aarch64) instead of golang arches (e.g. arm64).
+        dest_image_tag = f"{release_name}-{brew_arch}"
+        dest_image_pullspec = f"{self.DEST_RELEASE_IMAGE_REPO}:{dest_image_tag}"
+        self._logger.info("Checking if release image %s for %s (%s) already exists...", release_name, arch, dest_image_pullspec)
```
Could this be a `f'string'`?

vfreex replied:
f'string in logging methods fails some lint tool checks.

joepvd commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+        self._logger.info("Checking if release image %s for %s (%s) already exists...", release_name, arch, dest_image_pullspec)
+        dest_image_info = await self.get_release_image_info(dest_image_pullspec)
+        if dest_image_info:  # this arch-specific release image is already promoted
+            self._logger.warning("Release image %s for %s (%s) already exists", release_name, arch, dest_image_info["image"])
```
f-string

joepvd commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+        dest_image_info = await self.get_release_image_info(dest_image_pullspec)
+        if dest_image_info:  # this arch-specific release image is already promoted
+            self._logger.warning("Release image %s for %s (%s) already exists", release_name, arch, dest_image_info["image"])
+            # TODO: Check if the existing release image matches the assembly definition.
```
This would be great! Maybe we should capture it as a card.

joepvd commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+            self._elliott_env_vars["ELLIOTT_DATA_PATH"] = self._ocp_build_data_url
+            self._doozer_env_vars["DOOZER_DATA_PATH"] = self._ocp_build_data_url
+
+    async def run(self):
```
This method is a tad long. Would love to see it split up in more functions, and `run()` is calling these functions in sequence. This would help maintaining an overview for the reader, as well as give entrypoints for testing.
Not a blocker for merging :)

vfreex replied:
It becomes longer as I was adding more functionalities to this job :)

joepvd reviewed (approved):
/lgtm

Left quite some nits, none of them blockers. I'm sure to get at some of them when I need to alter these code paths.

vfreex commented on pyartcd/pyartcd/pipelines/promote.py:
```diff
@@ -0,0 +1,628 @@
+        if release_config is None:
+            raise ValueError(f"To promote this release, assembly {self.assembly} must be explictly defined in releases.yml.")
+
+        release_issue_waivers = release_config.get("assembly", {}).get("waives", [])
```
Sorry I forgot to mention it. It is used to replace the active prompts in current promote job. I will update the docs and start a discussion soon.

vfreex commented on jobs/build/promote-assembly/Jenkinsfile:
```diff
@@ -0,0 +1,358 @@
+    def justifications =release_info.justifications ?: []
+    def quay_url = "quay.io/openshift-release-dev/ocp-release"
+    def client_type = 'ocp'
+    if (release_info.type == "candidate" && !release_info.assembly.startsWith('rc.') || release_info.type == "custom") {
```
Added comment here.

vfreex commented:
@joepvd Thank you for the review. I've addressed the issues that are easy to fix and updated this PR. Other suggestions (like unit test improvements and splitting `run` method) will go to future PRs.

joepvd commented:
/lgtm

🥳
